File: src/ProtocoloServidor.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE DeriveGeneric #-}

module ProtocoloServidor where

import Data.Aeson
import GHC.Generics (Generic)
import Data.Text (Text)
import Data.Aeson.Types (Parser)

type Usuario = Text
type Sala = Text
type Texto = Text
type Estado = Text
type Operacion = Text
type Resultado = Text
type Extra = Text

data MensajeServidor
  = Identificar { username :: Usuario }
  | CambiarEstado { status :: Estado }
  | SolicitarUsuarios
  | TextoPrivado { username :: Usuario, text :: Texto }
  | TextoPublico { text :: Texto }
  | CrearSala { roomname :: Sala }
  | Invitar { roomname :: Sala, usernames :: [Usuario] }
  | UnirseSala { roomname :: Sala }
  | UsuariosSala { roomname :: Sala }
  | TextoSala { roomname :: Sala, text :: Texto }
  | AbandonarSala { roomname :: Sala }
  | Desconectar
  deriving (Show, Generic)

data RespuestaServidor
  = Respuesta { operation :: Operacion, result :: Resultado, extra :: Extra }
  deriving (Show, Generic)

-- Validation Functions
validarLongitud :: Int -> Text -> Text -> Either String Text
validarLongitud maxLen nombre tipo
  | T.length nombre > maxLen = Left $ T.unpack tipo ++ " no puede exceder los " ++ show maxLen ++ " caracteres"
  | otherwise = Right nombre

validarNombreUsuario :: Text -> Either String Text
validarNombreUsuario = validarLongitud 8 "Usuario"

validarNombreCuarto :: Text -> Either String Text
validarNombreCuarto = validarLongitud 16 "Cuarto"

-- FromJSON Instances for decoding
instance FromJSON MensajeServidor where
  parseJSON = withObject "MensajeServidor" $ \v -> do
    tipoMsg <- v .: "type" :: Parser Text
    case tipoMsg of
      "IDENTIFY" -> do
        usuario <- v .: "username"
        case validarNombreUsuario usuario of
          Left err -> fail err
          Right usuarioValido -> return $ Identificar usuarioValido
      "NEW_ROOM" -> do
        cuarto <- v .: "roomname"
        case validarNombreCuarto cuarto of
          Left err -> fail err
          Right cuartoValido -> return $ CrearSala cuartoValido
      -- Other cases here
      _ -> fail $ "Unknown message type: " ++ T.unpack tipoMsg

File: Proyecto1-Chat/Python/Chat/mi_paquete/ProtocoloServidor.py
import json
from typing import Union, List, Dict, Any

Usuario = str
Sala = str
Texto = str
Estado = str
Operacion = str
Resultado = str
Extra = str

class MensajeServidor:
    def __init__(self, tipo: str, **kwargs):
        self.tipo = tipo
        for key, value in kwargs.items():
            setattr(self, key, value)

    @classmethod
    def identificar(cls, username: Usuario):
        return cls("IDENTIFY", username=username)

    @classmethod
    def cambiar_estado(cls, status: Estado):
        return cls("CHANGE_STATUS", status=status)

    @classmethod
    def solicitar_usuarios(cls):
        return cls("REQUEST_USERS")

    @classmethod
    def texto_privado(cls, username: Usuario, text: Texto):
        return cls("PRIVATE_TEXT", username=username, text=text)

    @classmethod
    def texto_publico(cls, text: Texto):
        return cls("PUBLIC_TEXT", text=text)

    @classmethod
    def crear_sala(cls, roomname: Sala):
        return cls("NEW_ROOM", roomname=roomname)

    @classmethod
    def invitar(cls, roomname: Sala, usernames: List[Usuario]):
        return cls("INVITE", roomname=roomname, usernames=usernames)

    @classmethod
    def unirse_sala(cls, roomname: Sala):
        return cls("JOIN_ROOM", roomname=roomname)

    @classmethod
    def usuarios_sala(cls, roomname: Sala):
        return cls("ROOM_USERS", roomname=roomname)

    @classmethod
    def texto_sala(cls, roomname: Sala, text: Texto):
        return cls("ROOM_TEXT", roomname=roomname, text=text)

    @classmethod
    def abandonar_sala(cls, roomname: Sala):
        return cls("LEAVE_ROOM", roomname=roomname)

    @classmethod
    def desconectar(cls):
        return cls("DISCONNECT")

class RespuestaServidor:
    def __init__(self, operation: Operacion, result: Resultado, extra: Extra):
        self.operation = operation
        self.result = result
        self.extra = extra

def validar_longitud(max_len: int, nombre: Texto, tipo: Texto) -> Union[str, Texto]:
    if len(nombre) > max_len:
        return f"{tipo} no puede exceder los {max_len} caracteres"
    return nombre

def validar_nombre_usuario(nombre: Texto) -> Union[str, Texto]:
    return validar_longitud(8, nombre, "Usuario")

def validar_nombre_cuarto(nombre: Texto) -> Union[str, Texto]:
    return validar_longitud(16, nombre, "Cuarto")

def from_json(data: Dict[str, Any]) -> MensajeServidor:
    tipo_msg = data.get("type")
    if tipo_msg == "IDENTIFY":
        usuario = data.get("username")
        validado = validar_nombre_usuario(usuario)
        if isinstance(validado, str):
            raise ValueError(validado)
        return MensajeServidor.identificar(validado)
    elif tipo_msg == "NEW_ROOM":
        cuarto = data.get("roomname")
        validado = validar_nombre_cuarto(cuarto)
        if isinstance(validado, str):
            raise ValueError(validado)
        return MensajeServidor.crear_sala(validado)
    # Other cases here
    else:
        raise ValueError(f"Unknown message type: {tipo_msg}")

